feat(linter): implement eslint(no-dupe-class-members)

YangchenYe323 opened the pull request:
Implement [no-dupe-class-members](https://eslint.org/docs/latest/rules/no-dupe-class-members)

Boshen commented:
Two points:

* I think the implementation is a bit over complicated - you can take a look at the eslint implementation. The hashmap comes down to something like

```rust
FxHashMap<(name, r#static), (Span, MethodDefinitionKind)>
```

And you can just check if there is a conflict on MethodDefinitionKind.

* make sure to take in https://typescript-eslint.io/rules/no-dupe-class-members/ as well

Boshen commented on crates/oxc_linter/src/rules/no_dupe_class_members.rs:
```diff
@@ -0,0 +1,195 @@
+);
+
+impl Rule for NoDupeClassMembers {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::Class(class) = node.get().kind() else { return; };
+
+        let mut property_table = PropertyTable::default();
```
This is the right way, because we need to check all the classes anyway.

One small improvement is to reserve capacity by the total number of class members.

YangchenYe323 commented:
> * I think the implementation is a bit over complicated - you can take a look at the eslint implementation. The hashmap comes down to something like
> 
> ```rust
> FxHashMap<(name, r#static), (Span, MethodDefinitionKind)>
> ```

I took this approach before. The last test case I added is why I switched to the current one:
```javascript
class A{get foo() {} set foo() {} get foo() {}}
```

The second `foo` will overwrite the first `foo` but the third foo is conflicting with the first one so it won't give the correct span. If we want to have two spans pointing at the duplicate member, we'd need to store a different span with each declared kind. Would it be simpler to use a `HashMap<(name, static), HashMap<Kind, Span>>` ? Or should we drop the two span diagnostics?



> * make sure to take in https://typescript-eslint.io/rules/no-dupe-class-members/ as well

Will do it.

Boshen commented:
> Would it be simpler to use a HashMap<(name, static), HashMap<Kind, Span>> ? Or should we drop the two span diagnostics?

Your implementation provides more info and context, let's go with your approach.

Boshen commented:
Awesome! Thank you!
